In [10]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import scipy as sci
from scipy import stats, special
import statsmodels.api as sm
import math
#import warnings
#warnings.filterwarnings('ignore')

Para estimar los parámetros que posee una determinada muestra(considerando una determinada distribución) existen variados métodos, en este caso se utilizará el método de la _Máxima Versimilitud_, el cual consiste en encontrar los valores máximos de la función de Verosimilitud.

$$ L(x|\theta) =  \prod_{i = 1}^{N} f(x_i|\theta)$$

Debido a la forma de esta, y aprovechando las propiedades de la función logaritmo, se suele utilizar la función _Log-Verosimilitud_, con la cual se obtiene el mismo resultado.

$$ l(x|\theta) = \sum_{i = 1}^{N} ln(f(x_i|\theta))$$

Para encontrar el máximo de la función _Log-Verosimilitud_ se utilizará el método de _Newton-Raphson_, un método iterativo el cual sirve para encontrar las raíces de una función real derivable. 

$$ x_{i + 1} = x_{i} + \frac{f(x_i)}{f'(x_i)}$$

Dado que encontrar el máximo de una función es equivalente a encontrar las raíces de su primera derivada, el método de Newton-Raphson quedaría de la siguiente manera.

$$ \theta_{i + 1} = \theta_{i} + \frac{l'(x|\theta_i)}{l''(x|\theta_i)}$$

Considerando una distribución Gamma con parámetro $\alpha$ desconocido y $\beta = 1$. 

$$ f(x) = \frac{1}{\Gamma(\alpha)} \cdot x^{\alpha - 1} \cdot e^{-x} $$

Al aplicar la función Log-Verosimilitud y derivar, se obtienen las primera dos derivadas necesarias para el método

$$l'(x|\alpha) = -N \cdot \frac{\Gamma'(\alpha)}{\Gamma(\alpha)} + \sum_{i = 1}^{N}ln(x_i)$$

$$l''(x|\alpha) = -N \cdot \left( \frac{\Gamma'(\alpha)}{\Gamma(\alpha)} \right)' $$

Con $x_i$ el elemento $i$ de la muestra a utilizar, $N$ la cantidad de elementos en la muestra y $\alpha$ el parámetro a utilizar.En este caso la muestra se encuentra en el archivo _"gamma.csv"_. 

Se programa ambas derivadas y se implementa el método de Newton-Raphson, considerando $\alpha_0 = 0.2$ y un umbral de detención de $u = 0.0000001$, es decir, el método se detendrá cuando la diferencia entre $\alpha_{i + 1}$ y $\alpha_{i}$ sea menor a dicho umbral.

In [16]:
def primeraDerivada(alpha, n, sumatoria):
    return (-n * special.digamma(alpha)) + sumatoria 
    
def segundaDerivada(alpha, n):
    return (-n * special.polygamma(1, alpha))

alpha = 0.2
newalpha = 0

umbral = 0.0000001

muestra = pd.read_csv('gamma.csv')
n = len(muestra['x'])
sumatoria = 0

for i in muestra['x']:
    sumatoria += math.log(i)

while(True):
    newalpha = alpha - (primeraDerivada(alpha, n, sumatoria)/segundaDerivada(alpha, n))
    if(abs(newalpha - alpha)< umbral):
        alpha = newalpha
        break
    alpha = newalpha
print(alpha)
    

2.60349215342


Aplicando el método, se obtiene que el estimador máximo verosimil de $\alpha$ es de 2.60349215342.